File: src/day18/solution.py
import sys
import os
import re
from collections import deque
with open(os.path.join(sys.path[0], 'input' if len(sys.argv)==1 else 'sample'), "r") as infile:
    lines = infile.read().split("\n")

def try_process(val_stack, op_stack):
    if(len(val_stack) >= 2 and len(op_stack)>=1):
        if(op_stack[-1] == "+" or op_stack[-1] == "*" ):
            t1 = val_stack.pop()
            t2 = val_stack.pop()
            res = t1+t2 if op_stack.pop()=="+" else t1*t2
            val_stack.append(res)


part1 = 0
part2 = 0
for line in lines:
    '''
    op_stack = deque()
    val_stack = deque()
    bracket_depth = 0
    symbols = list(line.replace(" ",""))
    for s in symbols:
        if(s == " "):
            continue
        elif(ord('0') <= ord(s) <= ord('9')):
            val_stack.append(int(s))
            try_process(val_stack, op_stack)
        elif(s == '+' or s == '*'):
            op_stack.append(s)
        elif(s == '('):
            bracket_depth += 1
            op_stack.append(s)
        elif(s == ')'):
            bracket_depth -= 1
            op_stack.pop()
            try_process(val_stack, op_stack)
    solution = val_stack.pop()
    part1 += solution
    '''
    symbols = list(line.replace(" ",""))
    rpn = []
    op_stack = deque(['$'])
    for s in symbols:
        if(ord('0') <= ord(s) <= ord('9')):
            rpn.append((0,int(s)))
        elif(s == '+' or s=='*'):
            while(op_stack[-1] != '$' and op_stack[-1]!='('):
                rpn.append((1, op_stack.pop()))
            op_stack.append(s)
        elif(s == '('):
            op_stack.append('(')
        elif(s == ')'):
            while(op_stack[-1]!='('):
                rpn.append((1, op_stack.pop()))
            op_stack.pop()
    while(op_stack[-1]!='$'):
        rpn.append((1, op_stack.pop()))
    
    
    acc = deque()
    for (code, val) in rpn:
        if(code == 0):
            acc.append(val)
        else:
            t1 = acc.pop()
            t2 = acc.pop()
            acc.append(t1+t2 if val=="+" else t1*t2)
    solution = acc[0]
    part1+=solution
    
    

    rpn = []
    op_stack = deque(['$'])
    for s in symbols:
        if(ord('0') <= ord(s) <= ord('9')):
            rpn.append((0,int(s)))
        elif(s == '+'):
            op_stack.append('+')
        elif(s == '*'):
            while(op_stack[-1]=='+'):
                rpn.append((1, op_stack.pop()))
            op_stack.append('*')
        elif(s == '('):
            op_stack.append('(')
        elif(s == ')'):
            while(op_stack[-1]!='('):
                rpn.append((1, op_stack.pop()))
            op_stack.pop()
    while(op_stack[-1]!='$'):
        rpn.append((1, op_stack.pop()))
    
    
    acc = deque()
    for (code, val) in rpn:
        if(code == 0):
            acc.append(val)
        else:
            t1 = acc.pop()
            t2 = acc.pop()
            acc.append(t1+t2 if val=="+" else t1*t2)
    solution = acc[0]
    
    part2 += solution

    
    

print(part1)    
print(part2)


        

File: src/day18/Solution.java
package day18;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.Deque;
import java.util.LinkedList;
import java.util.List;
import utils.AoCSolvable;
import utils.NoSolutionException;

public class Solution implements AoCSolvable {
  private static final boolean USE_SAMPLE = false;

  private static class TagPair{
    private int tag;
    private long i;
    private char c;

    private TagPair(long i){
      this.tag = 0;
      this.i = i;
    }

    private TagPair(char c){
      this.tag = 1;
      this.c = c;
    }
  }

  private final List<char[] >symbols;

  Solution(List<String> lines) throws NoSolutionException {
    symbols = new ArrayList<>();
    for (String line:lines){
      symbols.add(line.replace(" ", "").toCharArray());
    }
  }

  private static long evalRPN(List<TagPair> revPolish) throws NoSolutionException{
    Deque<Long> stack = new LinkedList<>();
    for(TagPair t:revPolish){
      if(t.tag == 0) stack.addFirst(t.i);
      else{
        Long t1 = stack.removeFirst();
        Long t2 = stack.removeFirst();
        if(t.c == '*') stack.addFirst(t1*t2);
        else if(t.c == '+') stack.addFirst(t1+t2);
        else throw new NoSolutionException();
      }
    }
    if(stack.size()!=1) throw new NoSolutionException();
    else return stack.removeFirst();
  }

  public String partOne() throws NoSolutionException {
    long solution = 0;
    for (char[] symbolArr:symbols){
      List<TagPair> revPolish = new ArrayList<>();
      Deque<Character> opStack = new LinkedList<>();
      opStack.addFirst('$');
      for (char c:symbolArr){
        switch(c){
          case '0': case '1': case '2': case '3': case '4':
          case '5': case '6': case '7': case '8': case '9':
            revPolish.add(new TagPair(Character.getNumericValue(c)));
            break;
          case '+': case '*':
            while(opStack.peekFirst() != '$' && opStack.peekFirst() != '('){
              revPolish.add(new TagPair(opStack.removeFirst()));
            }
            opStack.addFirst(c);
            break;
          case '(':
            opStack.addFirst(c);
            break;
          case ')':
            while(opStack.peekFirst() != '('){
              revPolish.add(new TagPair(opStack.removeFirst()));
            }
            opStack.removeFirst();
            break;
          default:
            throw new NoSolutionException();
        }
      }
      while(opStack.peekFirst() != '$') revPolish.add(new TagPair(opStack.removeFirst()));
      solution +=  evalRPN(revPolish);
    }
    return String.valueOf(solution);
  }

  public String partTwo() throws NoSolutionException {
    long solution = 0;
    for (char[] symbolArr:symbols){

      List<TagPair> revPolish = new ArrayList<>();
      Deque<Character> opStack = new LinkedList<>();
      opStack.addFirst('$');
      for (char c:symbolArr){
        switch(c){
          case '0': case '1': case '2': case '3': case '4':
          case '5': case '6': case '7': case '8': case '9':
            revPolish.add(new TagPair(Character.getNumericValue(c)));
            break;
          case '+':
            opStack.addFirst(c);
            break;
          case '*':
            while(opStack.peekFirst() == '+'){
              revPolish.add(new TagPair(opStack.removeFirst()));
            }
            opStack.addFirst(c);
            break;
          case '(':
            opStack.addFirst(c);
            break;
          case ')':
            while(opStack.peekFirst() != '('){
              revPolish.add(new TagPair(opStack.removeFirst()));
            }
            opStack.removeFirst();
            break;
          default:
            throw new NoSolutionException();
        }
      }
      while(opStack.peekFirst() != '$') revPolish.add(new TagPair(opStack.removeFirst()));
      solution += evalRPN(revPolish);
    }
    return String.valueOf(solution);
  }

  public static void main(String[] args) {
    try{
      List<String> lines = Files.readAllLines(Paths.get((USE_SAMPLE?"./sample/":"./input/")+Solution.class.getPackageName()));
      Solution s = new Solution(lines);
      System.out.println(s.partOne());
      System.out.println(s.partTwo());
    } catch (IOException | NoSolutionException e) {
      e.printStackTrace();
    }
  }
}
